src/schedule/zephyr_domain.c

```diff
@@ -76,7 +77,7 @@ static void zephyr_domain_thread_fn(void *p1, void *p2, void *p3)
 
 	for (;;) {
 		/* immediately go to sleep, waiting to be woken up by the timer */
-		k_thread_suspend(_current);
```

Comment:
Good catch.  Yeah, code like this where you try to suspend yourself directly and be woken up from somewhere else is inherently racy, as the decision you just made ("I should be suspended") is based on data that isn't inside any kind of lock.

Semaphore are the perfect primitive to use here.  But if you have more complicated usages note we just added a condition variable that allows you to have arbitrary state inside a mutex, and atomically "go to sleep and release the lock" in a single call.

Reply:
@andyross yes, I thought about using a condition variable first, but the "waker" has to hold a mutex to wake the suspended thread, and you don't want to take a mutex in the interrupt context...